### PR TITLE
Remove `cache_position` in more models

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title! Follow-up of https://github.com/huggingface/transformers/pull/44181 with more models!

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
run-slow: got_ocr2 marian lfm2_vl modernbert_decoder video_llama_3 qwen2_5_omni qwen3_omni_moe biogpt janus dbrx fast_vlm bart perception_lm blenderbot llava_onevision roberta gemma2 llava ovis2 deepseek_vl_hybrid mistral3 paligemma glm_ocr t5gemma2 ernie bigbird_pegasus pegasus t5gemma glm4v_moe time_series_transformer timesfm2_5 ernie4_5_moe voxtral llava_next_video nanochat qwen2_5_vl gpt_neox xlm_roberta_xl deepseek_vl vipllava cohere2 big_bird gemma3n granitemoehybrid internvl git lighton_ocr gpt_oss colqwen2 m2m_100 glm46v plbart aya_vision bamba metaclip_2 camembert zamba2 blt informer glm_image gemma3 mbart blenderbot_small cohere2_vision phi4_multimodal glm4v roberta_prelayernorm aimv2 smolvlm florence2 vaultgemma data2vec_text jetmoe pegasus_x

**Comment by Cyrilvallez** [comment]
run-slow: got_ocr2 marian lfm2_vl modernbert_decoder video_llama_3 qwen2_5_omni qwen3_omni_moe janus dbrx fast_vlm bart perception_lm blenderbot roberta gemma2 llava ovis2 deepseek_vl_hybrid mistral3 paligemma glm_ocr t5gemma2 pegasus t5gemma glm4v_moe time_series_transformer timesfm2_5 ernie4_5_moe voxtral llava_next_video nanochat qwen2_5_vl gemma3n

**Comment by Cyrilvallez** [comment]
run-slow: lfm2_vl modernbert_decoder qwen2_5_omni qwen3_omni_moe bart gemma2 llava gemma3n

**Comment by Cyrilvallez** [comment]
Because the `run-slow` command does not work, I checked locally and all those models: 
`
got_ocr2 marian lfm2_vl modernbert_decoder video_llama_3 qwen2_5_omni qwen3_omni_moe janus dbrx fast_vlm bart perception_lm blenderbot roberta gemma2 llava ovis2 deepseek_vl_hybrid mistral3 paligemma glm_ocr t5gemma2 pegasus t5gemma glm4v_moe time_series_transformer timesfm2_5 ernie4_5_moe voxtral llava_next_video nanochat qwen2_5_vl gemma3n
`
have no diffs with main on their IntegrationTest!
So all good!

File: src/transformers/models/bamba/modeling_bamba.py [hunk 1]
```diff
@@ -611,8 +610,6 @@ def cuda_kernels_forward(
             and cache_params.conv_states[self.layer_idx].shape[0]
             == cache_params.ssm_states[self.layer_idx].shape[0]
             == batch_size
-            and cache_position is not None
-            and cache_position[0] > 0
```
Comment (vasqu) [inline src/transformers/models/bamba/modeling_bamba.py:615]: Honestly we never needed that condition at all, given we have `cache_params.has_previous_state`

File: src/transformers/models/bamba/modeling_bamba.py [hunk 2]
```diff
@@ -1266,14 +1249,16 @@ def forward(
...
         """
         No need for zeroing states when
             1. Cached forward
             2. Attending to all inputs
         """
         mamba_mask = attention_mask
-        if cache_position[0] > 0 or (attention_mask is not None and torch.all(attention_mask == 1)):
+        if (past_key_values is not None and past_key_values.get_seq_length() > 0) or (
+            attention_mask is not None and torch.all(attention_mask == 1)
+        ):
```
Comment (vasqu) [inline src/transformers/models/bamba/modeling_bamba.py]:
```suggestion
        if (past_key_values is not None and past_key_values.has_previous_state) or (
            attention_mask is not None and torch.all(attention_mask == 1)
        ):
```
imo we could also do this which is more obvious to me, maybe that's just me

File: src/transformers/models/bamba/modeling_bamba.py
Comment (vasqu) [inline src/transformers/models/bamba/modeling_bamba.py:1]: Will likely be the same for most hybrid models so not commentin the same there

File: src/transformers/models/bart/modeling_bart.py [hunk 3]
```diff
@@ -786,10 +771,7 @@ def forward(
 
         batch_size, seq_length = inputs_embeds.size()[:-1]
         past_key_values_length = past_key_values.get_seq_length() if past_key_values is not None else 0
-        if cache_position is None:
-            cache_position = torch.arange(
-                past_key_values_length, past_key_values_length + seq_length, device=inputs_embeds.device
-            )
+        position_ids = torch.arange(seq_length, device=inputs_embeds.device) + past_key_values_length
```
Comment (vasqu) [inline src/transformers/models/bart/modeling_bart.py:774]: Position ids are always `None` I assume?

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 4]
```diff
@@ -428,13 +413,13 @@ def forward(
...
             inputs_embeds=inputs_embeds,
             attention_mask=attention_mask,
-            cache_position=cache_position,
             past_key_values=self_attn_cache,
         )
 
         # embed positions
         if position_ids is None:
-            position_ids = cache_position.unsqueeze(0)
+            position_ids = torch.arange(seq_length, device=inputs_embeds.device) + past_key_values_length
```
Comment (vasqu) [inline src/transformers/models/biogpt/modeling_biogpt.py:421]: Seeing it multiple times but is there any difference for torch to do `arange + int` vs `arange` only?

File: src/transformers/models/ernie/modeling_ernie.py [hunk 5]
```diff
@@ -644,19 +643,10 @@ def forward(
             raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
         elif input_ids is not None:
             self.warn_if_padding_and_no_attention_mask(input_ids, attention_mask)
-            input_shape = input_ids.size()
-        elif inputs_embeds is not None:
-            input_shape = inputs_embeds.size()[:-1]
-        else:
+        elif input_ids is None and inputs_embeds is None:
             raise ValueError("You have to specify either input_ids or inputs_embeds")
```
Comment (vasqu) [inline src/transformers/models/ernie/modeling_ernie.py]: Imo, we can directly reduce to 
```py
     if (input_ids is None) ^ (inputs_embeds is not None):
            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
```
only

File: src/transformers/models/ernie/modular_ernie.py [hunk 6]
```diff
@@ -290,22 +278,20 @@ def forward(
             past_key_values=encoder_outputs.past_key_values,
         )
 
-    # Copied from transformers.models.bert.modeling_bert.BertModel._create_attention_masks
+    # No longer Copied from transformers.models.bert.modeling_bert.BertModel._create_attention_masks
```
Comment (vasqu) [inline src/transformers/models/ernie/modular_ernie.py:277]: Hmm a bit weird to change this model but not bert 🤔 how come?

File: src/transformers/models/glm_image/modeling_glm_image.py [hunk 7]
```diff
@@ -1277,12 +1264,15 @@ def compute_3d_position_ids(
         elif self.rope_deltas is not None:
             batch_size, seq_length, _ = inputs_embeds.shape
             if self._cached_decode_position_ids is not None:
-                step = cache_position[0].item() - self._prefill_len
+                step = past_key_values_length - self._prefill_len
```
Comment (vasqu) [inline src/transformers/models/glm_image/modeling_glm_image.py:1267]: Interesting, could've been done before as well without the device sync - I guess you already double checked for correctness

File: src/transformers/models/modernbert_decoder/modeling_modernbert_decoder.py [hunk 8]
```diff
@@ -475,35 +470,26 @@ def forward(
...
         **kwargs: Unpack[TransformersKwargs],
     ) -> tuple[torch.Tensor, ...] | BaseModelOutputWithPast:
         if (input_ids is None) == (inputs_embeds is None):
             raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
 
         if input_ids is not None:
             self.warn_if_padding_and_no_attention_mask(input_ids, attention_mask)
-            batch_size, seq_length = input_ids.shape[:2]
-        else:
-            batch_size, seq_length = inputs_embeds.shape[:2]
```
Comment (vasqu) [inline src/transformers/models/modernbert_decoder/modeling_modernbert_decoder.py]: Same comment to reduce this to the ^ version only then

**Review by vasqu — APPROVED** [review]
Only thing I would address
- Hybrid models nit
- Some simplification on the input_ids/embeds check on a small number of models (iirc 2 I commented on)

The rest is mostly me commenting to myself a bit or nits

File: src/transformers/models/bamba/modeling_bamba.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/bamba/modeling_bamba.py:615]: Yup, no idea why it was there... But that's why I fully removed!

File: src/transformers/models/bart/modeling_bart.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/models/bart/modeling_bart.py:774]: Yup, since for some reason later on it was doing `positions = self.embed_positions(..., position_ids=cache_position)` and cache_position can never be different from a very simple `arange` based on the current length (i.e. it does not have the subtlety of `position_ids` which are different based on padding, packing format etc...), it's the same to recreate the `arange` in this way! So independently of if the "real" `position_ids` are passed as kwarg. This is because the model recreates the positions anyway

File: src/transformers/models/biogpt/modeling_biogpt.py (hunk 4 above)
Comment (Cyrilvallez) [reply src/transformers/models/biogpt/modeling_biogpt.py:421]: Yes! It's a subtlety, but because `past_key_values_length` will be a tensor if and only if the cache is a StaticCache, doing it this way avoids a precious cuda synchronization for those cases! See https://github.com/huggingface/transformers/pull/42433 and the issue mentioned for more details!

File: src/transformers/models/ernie/modular_ernie.py (hunk 6 above)
Comment (Cyrilvallez) [reply src/transformers/models/ernie/modular_ernie.py:277]: Was focusing on finishing the models using modular, and a LOT of other models are copied from bert, so will keep it for next batch!
